ext/drstatecmp/drstatecmp.c

```diff
@@ -0,0 +1,572 @@
+
+/* Label types. */
+typedef enum {
+    DRSTATECMP_LABEL_TERM,    /* Denotes the terminator of the original bb. */
+    DRSTATECMP_LABEL_ORIG_BB, /* Denotes the original bb. */
+    DRSTATECMP_LABEL_COPY_BB, /* Denotes the bb copy. */
+    DRSTATECMP_LABEL_COUNT,
+} drstatecmp_label_t;
+
+/* Reserve space for the label values. */
+static ptr_uint_t label_base;
+static void
+drstatecmp_label_init(void)
+{
+    label_base = drmgr_reserve_note_range(DRSTATECMP_LABEL_COUNT);
+    ASSERT(label_base != DRMGR_NOTE_NONE, "failed to reserve note space");
+}
+
+/* Get label values. */
+static inline ptr_int_t
+get_label_val(drstatecmp_label_t label_type)
+{
+    return (ptr_int_t)(label_base + label_type);
+}
+
+/* Compare instr label against a label_type. Returns true if they match. */
+static inline bool
+match_label_val(instr_t *instr, drstatecmp_label_t label_type)
+{
+    return instr_get_note(instr) == (void *)get_label_val(label_type);
+}
+
+/* Create and insert bb labels. */
+static instr_t *
+drstatecmp_insert_labels(void *drcontext, instrlist_t *ilist, instr_t *where,
+                         drstatecmp_label_t label_type, bool preinsert)
+{
+    instr_t *label = INSTR_CREATE_label(drcontext);
+    instr_set_meta(label);
+    instr_set_note(label, (void *)get_label_val(label_type));
+    if (preinsert)
+        instrlist_meta_preinsert(ilist, where, label);
+    else
+        instrlist_meta_postinsert(ilist, where, label);
+    return label;
+}
+
+/* Returns whether or not instr may have side effects. */
+static bool
+drstatecmp_may_have_side_effects_instr(instr_t *instr)
+{
+    /* Instructions with side effects include instructions that write to memory,
+     * interrupts, and syscalls. To avoid inter-procedural analysis, all function calls
+     * are conservatively assumed to have side effects.
+     */
+    return instr_writes_memory(instr) || instr_is_interrupt(instr) ||
+        instr_is_syscall(instr) || instr_is_call(instr);
+}
+
+/****************************************************************************
+ * APPLICATION-TO-APPLICATION PHASE
+ *
+ * In this phase, the side-effect-free basic blocks are duplicated. */
+
+/* Duplicates the bb if it is side-effect-free. */
+static dr_emit_flags_t
+drstatecmp_duplicate_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+                           bool translating, OUT void **user_data)
+{
+    /* Allocate space for user_data. */
+    drstatecmp_user_data_t *data = (drstatecmp_user_data_t *)dr_thread_alloc(
+        drcontext, sizeof(drstatecmp_user_data_t));
+    memset(data, 0, sizeof(*data));
+    *user_data = (void *)data;
+
+    /* Determine whether the basic block is free of side effects. */
+    bool side_effect_free = true;
+    for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+         inst = instr_get_next_app(inst)) {
+        side_effect_free &= !drstatecmp_may_have_side_effects_instr(inst);
```

Comment:
Avoid bitwise operators for bools, especially in C where there's no true bool type: safest to use logical &&

Reply:
Done.